### PR TITLE
Use allegrex-enabled GNU AS for PSP

### PR DESCRIPTION
**Pull request by sozud**
This uses a pre-compiled version of binutils 2.42 with allegrex support from here: https://github.com/sozud/binutils-allegrex
The script shows how it's built. The tar archive includes other tools but we can probably trim it down to just AS.

### CODE REVIEW

**Comment by bismurphy** [comment]
Could this be adjusted so that it builds through Github? Only reason I ask is that distributing a binary is a little sketchy, since you could theoretically change the executable to be something else, and if we're just cloning an executable from your repo, there's nothing to connect it to the real code from binutils. I'm not saying that you personally would do this, as I have every reason to trust in your history with this project; just as a general security rule it would be good if we could make sure that the executable has a chain of history to connect it with known code.

**Comment by Xeeynamo** [comment]
Shall we just add the dependency here https://github.com/xeeynamo/sotn-decomp/releases ? It would be integrated like any other binary we put in `bin/`. Not sure if you have permission to edit that. If not I can do it

**Comment by sozud** [comment]
OK, I uploaded the artifact to releases. 

What triggers this?

```
bin/%: bin/%.tar.gz
	sha256sum --check $<.sha256
	cd bin && tar -xzf ../$<
	rm $<
	touch $@
bin/%.tar.gz: bin/%.tar.gz.sha256
	wget -O $@ https://github.com/Xeeynamo/sotn-decomp/releases/download/cc1-psx-26/$*.tar.gz
```
It doesn't seem to be running for the psp build.

**Comment by Xeeynamo** [comment]
> OK, I uploaded the artifact to releases.
> 
> What triggers this?
> 
> ```
> bin/%: bin/%.tar.gz
> 	sha256sum --check $<.sha256
> 	cd bin && tar -xzf ../$<
> 	rm $<
> 	touch $@
> bin/%.tar.gz: bin/%.tar.gz.sha256
> 	wget -O $@ https://github.com/Xeeynamo/sotn-decomp/releases/download/cc1-psx-26/$*.tar.gz
> ```
> 
> It doesn't seem to be running for the psp build.

Something like this: https://github.com/Xeeynamo/sotn-decomp/pull/1257/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R123

so if you do `make bin/allegrex` it will download, check the sum and deflate it. 

So in `$(PSP_BUILD_DIR)/asm/psp%.s.o: asm/psp%.s` you just add `bin/allegrex` like the following:

`$(PSP_BUILD_DIR)/asm/psp%.s.o: asm/psp%.s bin/allegrex`

Another question is how we could instruct mwccgap to acknowledge the new compiled assembler, as it is currently hardcoded into the script: https://github.com/mkst/mwccgap/blob/main/mwccgap/mwccgap.py#L19

File: Makefile.psp.mk [hunk 1]
```diff
@@ -47,10 +48,10 @@ $(MWCCGAP_APP):
 
 $(PSP_BUILD_DIR)/%.c.o: %.c $(MWCCPSP) $(MWCCGAP_APP)
```
Comment (Xeeynamo) [inline Makefile.psp.mk:49]:
```suggestion
$(PSP_BUILD_DIR)/%.c.o: %.c $(MWCCPSP) $(MWCCGAP_APP) bin/allegrex-as
```

**Comment by Xeeynamo** [comment]
Try adding this:
```
bin/allegrex-as: bin/allegrex
```
or just change `bin/allegrex-as` to `bin/allegrex`. Otherwise the makefile will try looking for `bin/allegrex-as.tar.gz`

**Comment by mkst** [comment]
If you rebase/pull once #1280 is merged you should get an :OK: build with the newer binutils version :+1:

**Comment by mkst** [comment]
```
Fatal error: can't create build/pspeu/asm/pspeu/st/wrp_psp/data/10260.data.s.o: No such file or directory
```
I guess there needs to be a mkdir somewhere which isn't happening with the change of assembler dependency?

**Comment by mkst** [comment]
Ah...
```
mipsel-linux-gnu-as -Iinclude -march=r3000 -mtune=r3000 -no-pad-sections -O1 -G0 -o build/pspeu/asm/pspeu/st/wrp_psp/data/10260.data.s.o asm/pspeu/st/wrp_psp/data/10260.data.s
mipsel-linux-gnu-as -Iinclude -march=r3000 -mtune=r3000 -no-pad-sections -O1 -G0 -o build/pspeu/asm/pspeu/st/wrp_psp/data/102C0.data.s.o asm/pspeu/st/wrp_psp/data/102C0.data.s
mipsel-linux-gnu-as -Iinclude -march=r3000 -mtune=r3000 -no-pad-sections -O1 -G0 -o build/pspeu/asm/pspeu/st/wrp_psp/data/[10](https://github.com/Xeeynamo/sotn-decomp/pull/1253/checks#step:11:11)4C0.data.s.o asm/pspeu/st/wrp_psp/data/104C0.data.s
mipsel-linux-gnu-as -Iinclude -march=r3000 -mtune=r3000 -no-pad-sections -O1 -G0 -o build/pspeu/asm/pspeu/st/wrp_psp/data/106C0.data.s.o asm/pspeu/st/wrp_psp/data/106C0.data.s
```
CI is using the rule from the Makefile:

```Makefile
$(BUILD_DIR)/%.s.o: %.s
	$(AS) $(AS_FLAGS) -o $@ $<
```
rather than the rule from the `Makefile.psp.mk`:
```Make
$(PSP_BUILD_DIR)/asm/psp%.s.o: asm/psp%.s
	mkdir -p $(dir $@)
	$(ASPSP) -o $@ $<
```

.. when I compile `master` locally it's using the PSP rule:
```
mipsel-linux-gnu-as -I include/ -G0 -march=r6000 -mabi=eabi -o build/pspeu/asm/pspeu/st/wrp_psp/data/10260.data.s.o asm/pspeu/st/wrp_psp/data/10260.data.s
```

**Comment by sozud** [comment]
I don't understand makefiles, giving up on this one

**Pull request closed without merging**